includes/Services/SiteGenService.php

```diff
@@ -253,6 +254,260 @@ public static function generate_child_theme( $data ) {
+		if (!is_array($color_palettes)) {
+			$color_palettes = json_decode($color_palettes, true);
+		}
+
+		// Retrieve the existing homepages to find the last version number.
+		$existing_homepages = get_option('nfd-sitegen-homepages', []);
+		$last_version_number = self::get_last_version_number($existing_homepages);
+		$version_number = $last_version_number + 1;
+
+		foreach ($home_pages as $key => $blocks) {
+			if (!is_array($blocks)) {
+				continue;
+			}
+
+			$filtered_blocks = array_filter($blocks, function($value) {
+				return !is_null($value);
+			});
+
+			$content = implode('', $filtered_blocks);
+
+			// Select a random palette and check against the parent's palette.
+			$palette_index = array_rand($color_palettes['colorpalette']);
+			$selected_palette = self::transform_palette($color_palettes['colorpalette'][$palette_index], $palette_index);
+
+			// If regeneration is true and the selected palette matches the parent's palette, reselect.
+			if ($regenerate && $regenerateColorPalettes) {
+				while ($selected_palette == $regenerateColorPalettes && count($color_palettes['colorpalette']) > 1) {
+					$palette_index = array_rand($color_palettes['colorpalette']);
+					$selected_palette = self::transform_palette($color_palettes['colorpalette'][$palette_index], $palette_index);
+				}
+			}
+
+			$version_info = [
+				"slug" => "version" . $version_number,
+				"title" => "Version " . $version_number,
+				"isFavourited" => false,
+				"content" => $content,
+				"color" => $selected_palette
+			];
+
+			$versions[] = $version_info;
+			$version_number++;
+		}
+
+		return $versions;
+	}
+
+	/**
+	 * processes the Homepages response structure for favourited prviews
+	 *
+	 * @return array
+	 */
+	public static function process_favorited_regenerate(
+		$home_pages,
+		$regenerateColorPalattes
+	){
+		$versions = [];
+		// Fetch the color palette data from the options table.
+		$color_palettes = get_option('nfd-ai-site-gen-colorpalette');
+
+		// Decode the color palettes if it's not an array (assuming it's a JSON string).
+		if (!is_array($color_palettes)) {
+			$color_palettes = json_decode($color_palettes, true);
+		}
+
+		// Retrieve the existing homepages to find the last version number.
+		$existing_homepages = get_option('nfd-sitegen-homepages', []);
+
+		// Select a random palette and check against the parent's palette.
+		$palette_index = array_rand($color_palettes['colorpalette']);
+		$selected_palette = self::transform_palette($color_palettes['colorpalette'][$palette_index], $palette_index);
+
+		// If regeneration is true and the selected palette matches the parent's palette, reselect.
+		if ($regenerateColorPalattes) {
+			while ($selected_palette == $regenerateColorPalattes && count($color_palettes['colorpalette']) > 1) {
+				$palette_index = array_rand($color_palettes['colorpalette']);
+				$selected_palette = self::transform_palette($color_palettes['colorpalette'][$palette_index], $palette_index);
+			}
+		}
```

Comment:
with while, it will avoid unnecessary execution of the loop when the first selected palette is already different from the palette(the version being regenerated).